benchmarks/k_means_kdd.py

```diff
@@ -0,0 +1,138 @@
+    parser.add_argument('-s', '--scheduler-address', default=None,
+                        help="Address for the scheduler node. Runs locally by "
+                             "default.")
+    parser.add_argument('--scikit-learn', action='store_true',
+                        help="Just use scikit-learn")
+
+    return parser.parse_args(args)
+
+
+def download():
+    p = os.path.join(get_data_home(), "kddcpu.data.gz")
+    if os.path.exists(p):
+        return p
+    r = requests.get(URL, stream=True)
+    with open(p, "wb") as f:
+        for chunk in r.iter_content(chunk_size=1024):
+            if chunk:
+                f.write(chunk)
+    return p
+
+
+def split(p):
+    output = os.path.join(get_data_home(), "kddcup.parq")
+    if not os.path.exists(output):
+
+        dtype = {
+            1: 'category',
+            2: 'category',
+            3: 'category',
+            41: 'category',
+        }
+
+        df = pd.read_csv(p, header=None, dtype=dtype)
+        cat_cols = df.select_dtypes(include=['category']).columns
+        df[cat_cols] = df[cat_cols].apply(lambda col: col.cat.codes)
+        df.columns = list(string.ascii_letters[:len(df.columns)])
+
+        ddf = dd.from_pandas(df, npartitions=16)
+        ddf.to_parquet(output)
+
+    return output
+
+
+def upload(p, fs):
+    fs.mkdir('dask-data/kddcup/kdd.parq')
+    for file in glob(os.path.join(p, '*')):
+        fs.put(file, 'dask-data/kddcup/kdd.parq/' + file.split('/')[-1])
+        print(file)
+
+
+def as_known(X, lengths):
+    blocks = X.to_delayed().flatten()
+    P = X.shape[1]
+    arrays = [
+        da.from_delayed(x, dtype=X.dtype, shape=(length, P))
+        for x, length in zip(blocks, lengths)
+    ]
+    return da.concatenate(arrays, axis=0)
+
+
+def run(path):
+    logger.info("Reading data")
+    df = dd.read_parquet(path)
+    df = df.persist()
+    logger.info("Data in memory")
+
+    # Get known chunks
+    lengths = df.map_partitions(len).compute()
+    X = as_known(df.values, lengths)
+
+    # Cluster
+    n_clusters = 8
+    oversampling_factor = 2
+    km = KMeans(n_clusters=n_clusters,
+                oversampling_factor=oversampling_factor,
+                random_state=0)
+    t0 = tic()
+    logger.info("Starting n_clusters=%2d, oversampling_factor=%2d",
+                n_clusters, oversampling_factor)
+    km.fit(X)
+    t1 = tic()
+    logger.info("Finished in %.2f", t1 - t0)
+
+
+def main(args=None):
+    args = parse_args(args)
+    logger.info("Checking local data")
+    local = split(download())
+
+    if args.scheduler_address:
+        logger.info("Using distributed mode")
+        client = Client(args.scheduler_address)
+        logger.info(client.scheduler_info())
+        fs = s3fs.S3FileSystem()
+        if fs.exists("dask-data"):
+            logger.info("Using cached dataset")
+        else:
+            logger.info("Uploading to cloud storage")
+            upload(local, fs)
+        path = "dask-data/kddcup/kdd.parq/"
```

Comment:
I tried running this with a dask cluster on localhost and ran into errors here

Reply:
Whoops, missing the `s3://`.

Some context, the kdd-cup dataset is the largest dataset used in the k-means|| paper, but it isn't that big... I can cluster the entire dataset just fine on my laptop.

I just setup a cluster to benchmark k-means on the airlines dataset.

Reply:
We shouldn't necessarily need to engage s3 here though, no?  I wanted to run this locally just to see the diagnostic dashboard (things look great by the way) and ran into issues here

Reply:
Oh I remember now. I assumed that using the distributed scheduler implied a remote cluster. I'll have to refactor it a bit more then.